packages/@aws-cdk/aws-glue/lib/connection.ts

```diff
@@ -0,0 +1,193 @@
+import * as ec2 from '@aws-cdk/aws-ec2';
+import * as cdk from '@aws-cdk/core';
+import * as constructs from 'constructs';
+import { CfnConnection } from './glue.generated';
+
+/**
+ * The type of the glue connection
+ *
+ * If you need to use a connection type that doesn't exist as a static member, you
+ * can instantiate a `ConnectionType` object, e.g: `new ConnectionType('NEW_TYPE')`.
+ */
+export class ConnectionType {
+  /** A list of all known `ConnectionType`s. */
+  public static readonly ALL = new Array<ConnectionType>();
```

Comment:
Why is this needed?

Reply:
Answering this comment and the following one about `ConnectionType`
- A few months back, `JDBC` was the only supported type
- When I started working on this PR, I did notice the new supported types
- I basically followed the pattern I noticed in https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-lambda/lib/runtime.ts#L39

Having said all of that, I'm happy to either
1. Switch to a normal enum (am not sure when is the next time they'll change or how frequent it is) or
1. Make all private

What do you think?

Reply:
I looked further - the glue cloudformation documentation is not super great on this area to be honest - and found mentions of even more types in https://docs.aws.amazon.com/glue/latest/dg/aws-glue-programming-etl-connect.html

This makes me a bit inclined to keeping it extensible rather than an enum

Reply:
I agree the documentation is not great. and I also agree it seems to be a more open system than typical enums. The documentation you referred to made me consider a different approach though. 

Since every connection type comes with its own connection options, it would be great to provide compile time assistance for that. Basically, we provide type-safe connection types, as well as raw connections as a sort of escape hatch:

```ts
// type safe API. type specification is not necessary, and all options are modeled and enforced according to
// https://docs.aws.amazon.com/glue/latest/dg/aws-glue-programming-etl-connect.html#aws-glue-programming-etl-connect-kafka
const kafkaConnection = new glue.KafkaConnection(this, 'KafkaConnection', {
  bootstrapServers: ['http://server1.kafka'],
  topicName: 'TheTopic',
  ...
});

// even type-safe API's will have an escape hatch:
kafkaConnection.addOption('key', 'value');

// escape hatch, free for all.
const parquetConnection = new glue.Connection(this, 'ParquetConnection', {
  type: 'parquet', // string as is, no enums here. 
  options: { // any key value 
  }
})
```

Would you be up to doing 2-3 of these type safe connections? I know it might be a little tedious :), but once we have that pattern locked-down, it will be easy to incrementally add type-safe API's by the community.

WDYT?

Reply:
Wait till you see the `Job` construct because I keep feeling it does require way more compile-time assistance.
I shied away from doing that for the size of work required to be honest and I kept thinking can we do something now that we can build on later without breaking the APIs. Let's continue that `Job` discussion for when we get there.

For this `Connection` work though, let me have a think about it. Maybe we separate it into ensuring we have the escape hatch version first and separate the work to add more compile-type safe versions on top.

I'll have a look and see what I can do and report back

Reply:
So, I have a couple of thoughts/questions
The main changes to support your suggestion would be to
1. Introduce new concrete classes with new props that are more geared towards the connection types
2. Be able to map their options into the proper options names
3. Provide that `addOption` or `addProperty` escape hatch to be able to add more things

In my mind, this can be achieved basically by extending the current `Connection`... maybe extract some `BaseConnectionProps` that drops `properties` in favour of each connection's options.

What I am not clear about - a pattern I haven't coded before - is the escape hatch... it basically means that the base resource cannot be created right away in the constructor, right? are there relevant methods/hooks to use to ensure that the base resource gets instantiated/synthesized after all escape hatch calls `addOption` are done?

Reply:
@humanzz 

> What I am not clear about - a pattern I haven't coded before - is the escape hatch... it basically means that the base resource cannot be created right away in the constructor, right? are there relevant methods/hooks to use to ensure that the base resource gets instantiated/synthesized after all escape hatch calls `addOption` are done?

The way we do this is by using `Lazy` values. It basically means that instead of assigning a concrete value to the underlying CFN property, you assign a function, that gets evaluated only at synth time. 

For example, you instantiate the underlying resource in the constructor as usual:

https://github.com/aws/aws-cdk/blob/9c5000dc3e0a750f43ffa620b3bef60c694387fc/packages/%40aws-cdk/aws-rds/lib/parameter-group.ts#L139-L143

Notice the `parameters` property is assigned with a `Lazy.any` that accepts a function, delaying the access to `this.parameters`.

This way - post instantiation additions are supported as well:

https://github.com/aws/aws-cdk/blob/9c5000dc3e0a750f43ffa620b3bef60c694387fc/packages/%40aws-cdk/aws-rds/lib/parameter-group.ts#L170-L173

For the sake of speediness though, I'm also perfectly ok with skipping the `addOption` method for now and just providing a `custom` key in the constructor:

```ts
const kafkaConnection = new glue.KafkaConnection(this, 'KafkaConnection', {
  topicName: 'MyTopic',
  custom: {
    'key': 'value',
  }
})
```

Whatever you feel you have capacity for.

Reply:
Nice... So I made some changes to the `Connection` and its props. I've also added an example `MongoDBConnection` class.
I don't feel comfortable adding other types coz testing that they actually do work and are passing the correct params would be quite time consuming because one has to really fill in the documentation gaps and test and iterate.

So, how about we go with what we have, if we think the pattern is correct and allows non-breaking future extensibility, and leave adding more type-safe implementations for whoever will be needing them and having more capacity to contribute?